### PR TITLE
pybind/mgr/selftest: add "mgr self-test eval" command

### PR DESCRIPTION
**Pull request by tchaikov**
and a simple REPL client allowing developer to peek and poke the
selftest module. if this turns out to be useful, we can promote this
method into a dedicated mix-in class, so other module can use it if
developer wants to test it manually.

Signed-off-by: Kefu Chai <kchai@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
Jenkins test make check

**Comment by tchaikov** [comment]
jenkins test api

**Comment by ronen-fr** [comment]
A useful idea. LGTM

File: src/pybind/ceph_mgr_repl.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/python3
+# -*- mode:python -*-
+# vim: ts=4 sw=4 smarttab expandtab
+
+__all__ = ['ConsoleOptions', 'MgrModuleInterpreter']
```
Comment (tchaikov) [inline src/pybind/ceph_mgr_repl.py:5]: in case some of us wants to reuse this python module.

**Review by epuertat — COMMENTED** [review]
Looks good @tchaikov , nothing to object about coding here. Perhaps adding some documentation and examples would help better understand what's this for and how to use it from CLI and from other modules.

I tried out and printed some internal vars, but I feel that I'm missing how to get the most of it.

Regarding the `repl` part... TBH I had to Google it. Are we already using that acronym in the Ceph project? If not, I'd suggest using something more intuitive, like `eval`...

File: src/pybind/mgr/selftest/module.py [hunk 2]
```diff
@@ -464,6 +468,25 @@ def _command_spam(self) -> None:
         self._event.clear()
         self.log.info("Ended command_spam workload...")
 
+    @CLICommand('mgr self-test repl')
+    def repl(self, s: str) -> HandleCommandResult:
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/mgr/selftest/module.py:472]: @epuertat this command is not supposed to be used directly. it should be used along with its REPL client.

File: src/pybind/ceph_mgr_repl.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                            ['name', 'conffile', 'prefix', 'timeout'])
+
+
+class MgrModuleInteractiveConsole(InteractiveConsole):
+    def __init__(self, opt, filename="<console>"):
+        super().__init__(filename)
+        self.cmd_prefix = opt.prefix
+        self.timeout = opt.timeout
+        self.cluster = rados.Rados(name=opt.name,
+                                   conffile=opt.conffile)
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/ceph_mgr_repl.py:47]: @epuertat `mgr.rados` is the one living in mgr, not in the client side.

**Comment by tchaikov** [comment]
> Looks good @tchaikov , nothing to object about coding here. Perhaps adding some documentation and examples would help better understand what's this for and how to use it from CLI and from other modules.
> 
> I tried out and printed some internal vars, but I feel that I'm missing how to get the most of it.

not sure how you interacted with it. if i were you, i'd just run

```console
$ ../src/pybind/ceph_mgr_repl.py
```

> 
> Regarding the `repl` part... TBH I had to Google it. Are we already using that acronym in the Ceph project? If not, I'd suggest using something more intuitive, like `eval`...

IMHO, REPL is a widely used acronym in computer science. we are in a REPL every day. and i encourage user to this tool as an interactive shell. that's why i named it repl. if you still think i should rename it. i will do. just wanted to emphasis the reason why i love this name.

**Comment by epuertat** [comment]
> > Looks good @tchaikov , nothing to object about coding here. Perhaps adding some documentation and examples would help better understand what's this for and how to use it from CLI and from other modules.
> > I tried out and printed some internal vars, but I feel that I'm missing how to get the most of it.
> 
> not sure how you interacted with it. if i were you, i'd just run
> 
> ```
> $ ../src/pybind/ceph_mgr_repl.py
> ```
> 
> in which, one can profile the different commands, poke the internals of a mgr, even debug a mgr module.

@tchaikov ah, ok. I just invoked the `mgr self-test repl <...>` one.

I tried that, but I don't get any information there:
```bash
[root@ceph ceph]# src/pybind/ceph_mgr_repl.py
Python 3.6.8 (default, Aug 24 2020, 17:57:11) 
[GCC 8.3.1 20191121 (Red Hat 8.3.1-5)] on linux
type "help", "copyright", "credits" or "license" for more information.
(MgrModuleInteractiveConsole)
[mgr self-test repl] >>> help()
the input is not complete[mgr self-test repl] >>>
```

> > Regarding the `repl` part... TBH I had to Google it. Are we already using that acronym in the Ceph project? If not, I'd suggest using something more intuitive, like `eval`...
> 
> IMHO, REPL is a widely used acronym in computer science. we are in a REPL every day. and i encourage user to use this tool in an interactive shell. that's why i named it repl. if you still think i should rename it. i will do. just wanted to emphasis the reason why i love this name.

If that's the case, no need of.

**Comment by tchaikov** [comment]
> > > Looks good @tchaikov , nothing to object about coding here. Perhaps adding some documentation and examples would help better understand what's this for and how to use it from CLI and from other modules.
> > > I tried out and printed some internal vars, but I feel that I'm missing how to get the most of it.
> > 
> > 
> > not sure how you interacted with it. if i were you, i'd just run
> > ```
> > $ ../src/pybind/ceph_mgr_repl.py
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > in which, one can profile the different commands, poke the internals of a mgr, even debug a mgr module.
> 
> @tchaikov ah, ok. I just invoked the `mgr self-test repl <...>` one.
> 
> I tried that, but I don't get any information there:
> 
> ```shell
> [root@ceph ceph]# src/pybind/ceph_mgr_repl.py
> Python 3.6.8 (default, Aug 24 2020, 17:57:11) 
> [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)] on linux
> type "help", "copyright", "credits" or "license" for more information.
> (MgrModuleInteractiveConsole)
> [mgr self-test repl] >>> help()
> the input is not complete[mgr self-test repl] >>>
> ```


```console
$ export LD_LIBRARY_PATH=/var/ssd/ceph/build/lib:$LD_LIBRARY_PATH
$ export PYTHONPATH=/var/ssd/ceph/src/pybind:/var/ssd/ceph/build/lib/cython_modules/lib.3:/var/ssd/ceph/src/python-common:$PYTHONPATH
$ ../src/pybind/ceph_mgr_repl.py
Python 3.9.2 (default, Feb 28 2021, 17:03:44)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
(MgrModuleInteractiveConsole)
[mgr self-test eval] >>> help()
Welcome to Python 3.9's help utility!
If this is your first time using Python, you should definitely check out
the tutorial on the Internet at https://docs.python.org/3.9/tutorial/.
Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To quit this help utility and
return to the interpreter, just type "quit".
To get a list of available modules, keywords, symbols, or topics, type
"modules", "keywords", "symbols", or "topics".  Each module also comes
with a one-line summary of what it does; to list the modules whose name
or summary contain a given string such as "spam", type "modules spam".
help>
You are now leaving help and returning to the Python interpreter.
If you want to ask for help on a particular object directly from the
interpreter, you can type "help(object)".  Executing "help('string')"
has the same effect as typing a particular string at the help> prompt.
[mgr self-test eval] >>> print(mgr.version)
ceph version Development (no_version) quincy (dev)
[mgr self-test eval] >>>
```

> 
>  > > Regarding the `repl` part... TBH I had to Google it. Are we already using that acronym in the Ceph project? If not, I'd suggest using something more intuitive, like `eval`...
> > 
> > 
> > IMHO, REPL is a widely used acronym in computer science. we are in a REPL every day. and i encourage user to use this tool in an interactive shell. that's why i named it repl. if you still think i should rename it. i will do. just wanted to emphasis the reason why i love this name.
> 
> If that's the case, no need of.

**Comment by tchaikov** [comment]
@epuertat i added a command to display instructions setting environment variables. could you take another look?

**Comment by tchaikov** [comment]
jenkins test api

**Review by epuertat — APPROVED** [review]
@tchaikov looks good to me, but I still think we should document this somewhere (probably a new section under the "Ceph Manager Daemon" one, something like "Debugging Ceph Manager...").

**Comment by tchaikov** [comment]
@epuertat thank you for your review. agreed. i will write one as a follow-up PR.

**Comment by tchaikov** [comment]
@epuertat see #41627

**Pull request merged**